src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,531 @@
+            : this(credential, properties, new DefaultRandomByteGenerator(), SystemClock.Instance)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="GssapiAuthenticator"/> class.
+        /// </summary>
+        /// <param name="username">The username.</param>
+        /// <param name="properties">The properties.</param>
+        public MongoAWSAuthenticator(string username, IEnumerable<KeyValuePair<string, string>> properties)
+            : this(username, properties, new DefaultRandomByteGenerator(), SystemClock.Instance)
+        {
+        }
+
+        internal MongoAWSAuthenticator(
+            UsernamePasswordCredential credential,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock dateTimeProvider)
+            : base(CreateMechanism(credential, properties, randomByteGenerator, dateTimeProvider))
+        {
+        }
+
+        internal MongoAWSAuthenticator(
+            string username,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock dateTimeProvider)
+            : base(CreateMechanism(username, null, properties, randomByteGenerator, dateTimeProvider))
+        {
+        }
+
+        /// <inheritdoc/>
+        public override string DatabaseName
+        {
+            get { return "$external"; }
+        }
+
+        // private static methods
+        private static MongoAWSMechanism CreateMechanism(
+            UsernamePasswordCredential credential,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock dateTimeProvider)
+        {
+            if (credential.Source != "$external")
+            {
+                throw new ArgumentException("MONGO AWS authentication may only use the $external source.", nameof(credential));
+            }
+
+            return CreateMechanism(credential.Username, credential.Password, properties, randomByteGenerator, dateTimeProvider);
+        }
+
+        private static MongoAWSMechanism CreateMechanism(
+            string username,
+            SecureString securePassword,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock dateTimeProvider)
+        {
+            var relativeUri = Environment.GetEnvironmentVariable("AWS_CONTAINER_CREDENTIALS_RELATIVE_URI");
+            var awsCredentialsCreators = new Func<AwsCredentials>[]
+            {
+                () => CreateAwsCredentialsFromMongoCredentials(username, securePassword, properties),
+                () => CreateAwsCredentialsFromEnvironmentVariables(),
+                () => CreateAwsCredentialsFromEcsResponse(relativeUri),
+                () => CreateAwsCredentialsFromEc2Response()
+            };
```

Comment:
This trick of having an array of creators seems overly complicated, Why can't this whole method just be:

    private static MongoAWSMechanism CreateMechanism(
        string username,
        SecureString securePassword,
        IEnumerable<KeyValuePair<string, string>> properties,
        IRandomByteGenerator randomByteGenerator,
        IClock clock)
    {
        var awsCredentials =
            CreateAwsCredentialsFromMongoCredentials(username, securePassword, properties) ??
            CreateAwsCredentialsFromEnvironmentVariables() ??
            CreateAwsCredentialsFromEcsResponse() ??
            CreateAwsCredentialsFromEc2Response();
        if (awsCredentials == null)
        {
            throw new ArgumentException("A MONGODB-AWS authenticator must have an access key ID.");
        }

        ValidateAwsCredentials(awsCredentials);

        var credentials = new UsernamePasswordCredential("$external", awsCredentials.Username, awsCredentials.Password);

        return new MongoAWSMechanism(credentials, awsCredentials.SessionToken, randomByteGenerator, clock);
    }

This might require some slight refactoring of the Create methods to only return null or a fully usable credential.

Reply:
I think it was my suggestion to have debugging a bit simpler. But I like Robert's idea too.